x/coins.go

```diff
@@ -262,3 +266,28 @@ func normalize(cs []*Coin) ([]*Coin, error) {
 
 	return coins, nil
 }
+
+// isNormalized check if coins are in normalized form. This is a cheap
+// operation.
+func isNormalized(cs []*Coin) bool {
+	var prev *Coin
+	for _, c := range cs {
+		if c == nil {
+			return false
+		}
+		if c.IsZero() {
```

Comment:
Instead of `c.IsZero()`, what about checking if `c.Validate()` returns an error?

It makes common sense checks, that this is a valid coin at all... otherwise we still have some assumptions that numbers are in range, etc. However, it is a bit more expensive operation.

Maybe benchmark the two and check if you like the approach. If so, you could use `cs.Validate()` instead of `isNormalized()`.

Otherwise, we assume all input is valid, just possibly in the wrong order? Can you check if Add, etc. handles input with out of range values? (It may well do so already and this function is good enough).

This approach may be the best, I just want a good argument why not to use existing Validate

Reply:
Adding additional `if c.Validate() != nil { return false }` does make quite a difference.

```benchmark                                           old ns/op     new ns/op     delta
BenchmarkCoinsNormalize/zero_len_coins-4            4.08          3.69          -9.56%
BenchmarkCoinsNormalize/one_coin-4                  5.94          6.33          +6.57%
BenchmarkCoinsNormalize/two_unordered_coins-4       20.5          302           +1373.17%
BenchmarkCoinsNormalize/four_normalized-4           33.7          1634          +4748.66%
BenchmarkCoinsNormalize/six_normalized-4            49.5          2069          +4079.80%
BenchmarkCoinsNormalize/twelve_normalized-4         95.6          5854          +6023.43%
BenchmarkCoinsNormalize/twelve_not_normalized-4     2561          2910          +13.63%
BenchmarkCoinsNormalize/nil_coins-4                 3.69          3.82          +3.52%
BenchmarkCoinsNormalize/two_normalized_coins-4      18.1          292           +1513.26%
BenchmarkCoinsNormalize/two_split_coins-4           185           487           +163.24%
BenchmarkCoinsNormalize/four_not_normalized-4       1054          1465          +38.99%
BenchmarkCoinsNormalize/six_not_normalized-4        1214          1510          +24.38%
```

Reply:
So the normalized performance goes horrible in this point, with the not-normalized part has a relatively minor hit.

> BenchmarkCoinsNormalize/four_normalized-4           33.7          1634          +4748.66%

I would just keep it was is, this is a good argument for not using the full Validate (I assume it is the regexp that is slow). Code from users should be `Validate`d anyway in another place before passing in, this should be more a check for mistakes...

Stick with your approach. I can approve this, once you remove the Validate function again. Thank you for running those benchmarks. And maybe add a comment about performance on those lines.

Also, you can use `c.IsEmpty()` to check for both nil and IsZero(), to make the code a bit shorter

Reply:
Nevermind, you never committed the Validate code. Smart man.